picard/util/imageinfo.py

```diff
@@ -43,189 +43,236 @@ class UnexpectedError(IdentificationError):
-        w = reader.bits(14) + 1
-        h = reader.bits(14) + 1
-    # Extended File Format
-    elif format == b'VP8X':
-        if len(data) < 30:
-            raise NotEnoughData('Not enough data for WebP VP8X')
-        reader = LSBBitReader(BytesIO(data[24:30]))
-        w = reader.bits(24) + 1
-        h = reader.bits(24) + 1
-    else:
-        h, w = 0, 0
-    return (w, h, 'image/webp', '.webp')
-
-
-TIFF_BYTE_ORDER_LSB = b'II'
-TIFF_BYTE_ORDER_MSB = b'MM'
-TIFF_TAG_IMAGE_LENGTH = 257
-TIFF_TAG_IMAGE_WIDTH = 256
-TIFF_TYPE_SHORT = 3
-TIFF_TYPE_LONG = 4
-
-
-def _tiff_read_value(type, order, data):
-    if type == TIFF_TYPE_LONG:
-        value = data[:4]
-        format = order + 'I'
-    elif type == TIFF_TYPE_SHORT:
-        value = data[:2]
-        format = order + 'H'
-    else:
-        raise UnexpectedError('TIFF: unexpected field type %s' % type)
-    return struct.unpack(format, value)[0]
-
-
-def _tiff_read(data):
-    # See https://www.adobe.io/content/dam/udp/en/open/standards/tiff/TIFF6.pdf
-    w, h = 0, 0
-    byte_order = data[:2]
-    if byte_order == TIFF_BYTE_ORDER_LSB:
-        order = '<'
-    elif byte_order == TIFF_BYTE_ORDER_MSB:
-        order = '>'
-    else:
-        raise UnexpectedError('TIFF: unexpected byte order %r' % byte_order)
-    try:
-        offset, = struct.unpack(order + 'I', data[4:8])
-        entry_count, = struct.unpack(order + 'H', data[offset:offset + 2])
-        pos = offset + 2
-        for i in range(entry_count):
-            field = data[pos:pos + 12]
-            tag, type = struct.unpack(order + 'HH', field[:4])
-            if tag == TIFF_TAG_IMAGE_WIDTH:
-                w = _tiff_read_value(type, order, field[8:12])
-            elif tag == TIFF_TAG_IMAGE_LENGTH:
-                h = _tiff_read_value(type, order, field[8:12])
-            if w and h:  # Found both width and height, abort
-                break
-            pos += 12
-    except struct.error:
-        pass
-    return (w, h, 'image/tiff', '.tiff')
-
-
-def _pdf_read(data):
-    return (0, 0, 'application/pdf', '.pdf')
+            self.h, self.w = 0, 0
+
+
+class IdentifyTiff(IdentifyImageType):
+    mime = 'image/tiff'
+    extension = '.tiff'
+
+    def match(self):
+        return self.data[:4] == b'II*\x00' or self.data[:4] == b'MM\x00*'
+
+    def _read(self):
+
+        TIFF_BYTE_ORDER_LSB = b'II'
+        TIFF_BYTE_ORDER_MSB = b'MM'
+        TIFF_TAG_IMAGE_LENGTH = 257
+        TIFF_TAG_IMAGE_WIDTH = 256
+        TIFF_TYPE_SHORT = 3
+        TIFF_TYPE_LONG = 4
+
+        def _tiff_read_value(type, order, data):
+            if type == TIFF_TYPE_LONG:
+                value = data[:4]
+                format = order + 'I'
+            elif type == TIFF_TYPE_SHORT:
+                value = data[:2]
+                format = order + 'H'
+            else:
+                raise UnexpectedError('TIFF: unexpected field type %s' % type)
+            return struct.unpack(format, value)[0]
```

Comment:
I would not define this inside the function but on the class itself (`self._read_value`, the tiff prefix can also be dropped).

Same probably for the constants.

Reply:
Agreed, done in 7bf8e06